### PR TITLE
check if flash unit is present

### PR DESCRIPTION
**Pull request by parveshneedhoo**
Added code to check if flash unit is present before toggling torch

### CODE REVIEW

File: src/android/SimpleCameraPreview.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import android.view.ViewGroup;
 import android.view.ViewParent;
 import android.widget.FrameLayout;
+import android.widget.Toast;
```
Comment (dinitri) [inline src/android/SimpleCameraPreview.java:21]:
```suggestion
```

File: src/android/CameraPreviewFragment.java [hunk 2]
```diff
@@ -224,6 +228,15 @@ public void torchSwitch(boolean torchOn, TorchCallback torchCallback) {
         }
     }
 
+    public void hasFlash(FlashCallback flashCallback) {
+        if (!camera.getCameraInfo().hasFlashUnit()) {
+            flashCallback.onResult(false);
+        } else {
+            flashCallback.onResult(true);
+        }
+        return;
```
Comment (dinitri) [inline src/android/CameraPreviewFragment.java:237]:
```suggestion
        flashCallback.onResult(camera.getCameraInfo().hasFlashUnit());
```
?

File: src/android/SimpleCameraPreview.java [hunk 3]
```diff
@@ -262,21 +265,36 @@ private boolean capture(boolean useFlash, CallbackContext callbackContext) {
...
-        callbackContext.error("Camera is closed, cannot switch " + torchState + " torch");
+    private boolean deviceHasFlash(CallbackContext callbackContext) {
+        fragment.hasFlash((boolean result) -> {
+            PluginResult pluginResult1;
+            if (result == true) {
+                pluginResult1 = new PluginResult(PluginResult.Status.OK, true);
+            } else {
+                pluginResult1 = new PluginResult(PluginResult.Status.OK, false);
+            }
+
```
Comment (dinitri) [inline src/android/SimpleCameraPreview.java:276]:
```suggestion
            PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, result);
```
can keep the name as pluginResult, it's a local variable

File: src/android/SimpleCameraPreview.java [hunk 4]
```diff
@@ -262,21 +265,36 @@ private boolean capture(boolean useFlash, CallbackContext callbackContext) {
...
+            pluginResult1.setKeepCallback(true);
```
Comment (dinitri) [inline src/android/SimpleCameraPreview.java:277]: why keep callback true ?

**Review by dinitri — CHANGES_REQUESTED** [review]
reviewed android part only

File: src/ios/SimpleCameraPreview.m [hunk 5]
```diff
@@ -132,6 +132,30 @@ - (void) torchSwitch:(CDVInvokedUrlCommand*)command{
...
+        }
+    }
+    
+    CDVPluginResult *pluginResult;
+    if (hasTorch == YES) {
+        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:true];
+    }
+    else {
+        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:false];
+    }
```
Comment (zfir) [inline src/ios/SimpleCameraPreview.m:155]:
```suggestion
      pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:hasTorch];
```

File: src/android/CameraPreviewFragment.java [hunk 6]
```diff
@@ -51,6 +51,10 @@ interface TorchCallback {
     void onEnabled(Exception err);
 }
 
+interface FlashCallback {
```
Comment (zfir) [inline src/android/CameraPreviewFragment.java:54]:
```suggestion
interface HasFlashCallback {
```
?

**Review by zfir — APPROVED** [review]

File: src/android/CameraPreviewFragment.java [hunk 7]
```diff
@@ -224,6 +228,11 @@ public void torchSwitch(boolean torchOn, TorchCallback torchCallback) {
         }
     }
 
+    public void hasFlash(HasFlashCallback flashCallback) {
```
Comment (zfir) [inline src/android/CameraPreviewFragment.java:231]:
```suggestion
    public void hasFlash(HasFlashCallback hasFlashCallback) {
```

**Review by zfir — APPROVED** [review]

File: src/ios/SimpleCameraPreview.m [hunk 8]
```diff
@@ -132,6 +132,24 @@ - (void) torchSwitch:(CDVInvokedUrlCommand*)command{
...
 
+- (void) deviceHasFlash:(CDVInvokedUrlCommand*)command{
+    AVCaptureDeviceDiscoverySession *captureDeviceDiscoverySession = [AVCaptureDeviceDiscoverySession discoverySessionWithDeviceTypes:@[AVCaptureDeviceTypeBuiltInWideAngleCamera]
+                                          mediaType:AVMediaTypeVideo
+                                           position:AVCaptureDevicePositionBack];
+    NSArray *captureDevices = [captureDeviceDiscoverySession devices];
+    BOOL hasTorch = NO;
+    
+    for (AVCaptureDevice *device in captureDevices) {
+        if ([device hasTorch]) {
```
Comment (dinitri) [inline src/ios/SimpleCameraPreview.m:143]: i see the hasTorch method too when looking at apple's documentation, should we cater for the hasTorch method too ?  or hasTorch is enough? @zfir just a question

**Review by dinitri — COMMENTED** [review]
just one question, it looks good

**Review by dinitri — APPROVED** [review]

**Pull request merged**